executor/executor_x.go

```diff
@@ -0,0 +1,341 @@
+			}
+			if e.subResult == nil {
+				return nil, nil
+			}
+		}
+		h, rowData, err := e.subResult.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if rowData == nil {
+			e.subResult = nil
+			continue
+		}
+		for i, field := range e.tablePlan.Fields() {
+			field.Expr.SetDatum(rowData[i])
+		}
+		return resultRowToRow(e.table, h, rowData), nil
+	}
+}
+
+// Fields implements Executor Fields interface.
+func (e *XSelectTableExec) Fields() []*ast.ResultField {
+	return e.tablePlan.Fields()
+}
+
+// Close implements Executor Close interface.
+func (e *XSelectTableExec) Close() error {
+	if e.result != nil {
+		e.result.Close()
+	}
+	return nil
+}
+
+func resultRowToRow(t table.Table, h int64, data []types.Datum) *Row {
+	entry := &RowKeyEntry{Handle: h, Tbl: t}
+	return &Row{Data: data, RowKeys: []*RowKeyEntry{entry}}
+}
+
+func (e *XSelectTableExec) doRequest() error {
+	txn, err := e.ctx.GetTxn(false)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	selReq := new(tipb.SelectRequest)
+	startTs := txn.StartTS()
+	selReq.StartTs = &startTs
+	selReq.Fields = resultFieldsToPBExpression(e.tablePlan.Fields())
+	selReq.Where = conditionsToPBExpression(e.tablePlan.FilterConditions...)
+	selReq.Ranges, selReq.Points = tableRangeToPBRangesAndPoints(e.tablePlan.Ranges)
+	selReq.TableInfo = tablecodec.TableToProto(e.tablePlan.Table)
+	e.result, err = xapi.Select(txn.GetClient(), selReq, 1)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// XSelectIndexExec represents XAPI select index executor.
+type XSelectIndexExec struct {
+	table     table.Table
+	indexPlan *plan.IndexScan
+	ctx       context.Context
+	where     *tipb.Expr
+	rows      []*Row
+	cursor    int
+}
+
+// Fields implements Executor Fields interface.
+func (e *XSelectIndexExec) Fields() []*ast.ResultField {
+	return e.indexPlan.Fields()
+}
+
+// Next implements Executor Next interface.
+func (e *XSelectIndexExec) Next() (*Row, error) {
+	if e.rows == nil {
+		err := e.doRequest()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	if e.cursor >= len(e.rows) {
+		return nil, nil
+	}
+	row := e.rows[e.cursor]
+	e.cursor++
+	return row, nil
+}
+
+// Close implements Executor Close interface.
+func (e *XSelectIndexExec) Close() error {
+	return nil
+}
+
+func (e *XSelectIndexExec) doRequest() error {
+	txn, err := e.ctx.GetTxn(false)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	idxResult, err := e.doIndexRequest(txn)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	handles, err := extractHandlesFromIndexResult(idxResult)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	indexOrder := make(map[int64]int)
+	for i, h := range handles {
+		indexOrder[h] = i
+	}
+	sort.Sort(int64Slice(handles))
+	tblResult, err := e.doTableRequest(txn, handles)
+	unorderedRows, err := extractRowsFromTableResult(e.table, tblResult)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Restore the original index order.
+	rows := make([]*Row, len(handles))
```

Comment:
index scan without order assumption is exceptional, this should be the default behavior.
We can add an option for out of order index scan later, but that require changes on optimizer.